_posts/2019-04-16-accessing-kafka-part-1.md

```diff
@@ -0,0 +1,107 @@
+
+In Strimzi, we currently support the second option.
+
+## Connecting from inside the same Kubernetes cluster
+
+Doing this for the clients running inside the same Kubernetes cluster as the Kafka cluster is running is actually quite simple.
+Each pod has its own IP address, which other applications can use to connect directly to it. 
+This is normally not used by regular Kubernetes applications.
+One of the reasons is that Kubernetes doesn't offer any nice way how to discover these IP addresses.
+To find out the IP address, you would need to use the Kubernetes API, find the right pod and its IP address.
+And you would need to have the rights for this.
+Instead, Kubernetes use the services with their stable DNS names as the main discovery mechanism.
+But with Kafka, this not an issue, because it has its own discovery protocol.
+We do not need the clients to figure out the API address from the Kubernetes API.
+We just need to configure it ad the advertised address and the clients will discover it through the Kafka _metadata_.
+
+But there is even one better option which is used by Strimzi.
+For StatefulSet - which Strimzi is using to run the Kafka brokers - you can use the Kubernetes headless service to give each of the pods a stable DNS name.
+Strimzi is using these DNS names as the advertised addresses for the Kafka brokers.
+So with Strimzi:
+* The initial connection is done using a regular Kubernetes service to get the _metadata_.
+* The subsequent connections are opened using the DNS names give to the pods by another headless Kubernetes service.
+The diagram below shows how does it look with an example Kafka cluster named `my-cluster`.
+
+![Accessing Kafka inside the same Kubernetes cluster]({{ "/assets/2019-04-16-inside-kubernetes.png" }})
+
+Both approaches have their own pros and cons.
+Using the DNS can sometimes cause problems with cached DNS information.
+When the underlying IP addresses of the pods change for example during rolling updates, the clients connecting to the brokers need to have the latest DNS information.
+However, we found out the using the IP addresses is causing even worse problems, because sometimes Kubernetes re-use them very aggressively and a new pod gets the IP address used just few seconds before by some other Kafka node.
+
+## Connecting from the outside
+
+While the access for clients running inside the same Kubernetes cluster was relatively simple, it will get a bit harder from the outside.
+While there are some tools for joining the Kubernetes network with the regular network outside of Kubernetes, most Kubernetes clusters run on their own network which is separated from the world outside.
+That means that the things such as pod IP addresses or DNS names are not resolvable for any clients running outside the cluster.
+Thanks to that, it is also clear that we will need to use separate Kafka listener for access from inside and outside of the cluster, because the advertised addresses will need to be different.
```

Comment:
```suggestion
Thanks to that, it is also clear that we will need to use a separate Kafka listener for access from inside and outside of the cluster, because the advertised addresses will need to be different.
```